internal/docstore/driver/codec.go

```diff
@@ -0,0 +1,701 @@
+// Copyright 2019 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO(jba): support struct tags.
+// TODO(jba): for efficiency, enable encoding of only a subset of field paths.
+
+package driver
+
+import (
+	"encoding"
+	"fmt"
+	"reflect"
+	"strconv"
+
+	"github.com/golang/protobuf/proto"
+	"gocloud.dev/internal/docstore/internal/fields"
+	"gocloud.dev/internal/gcerr"
+)
+
+var (
+	binaryMarshalerType   = reflect.TypeOf((*encoding.BinaryMarshaler)(nil)).Elem()
+	binaryUnmarshalerType = reflect.TypeOf((*encoding.BinaryUnmarshaler)(nil)).Elem()
+	textMarshalerType     = reflect.TypeOf((*encoding.TextMarshaler)(nil)).Elem()
+	textUnmarshalerType   = reflect.TypeOf((*encoding.TextUnmarshaler)(nil)).Elem()
+	protoMessageType      = reflect.TypeOf((*proto.Message)(nil)).Elem()
+)
+
+// An Encoder encodes Go values in some other form (e.g. JSON, protocol buffers).
+// The encoding protocol is designed to avoid losing type information by passing
+// values using interface{}. An Encoder is responsible for storing the value
+// it is encoding.
+//
+// Because all providers must support the same set of values, the encoding methods
+// (with the exception of EncodeStruct) do not return errors. EncodeStruct is special
+// because it is an escape hatch for arbitrary structs, not all of which may be
+// encodable.
+type Encoder interface {
+	// These methods all encode and store a single Go value.
+	EncodeNull()
+	EncodeBool(bool)
+	EncodeString(string)
+	EncodeInt(int64)
+	EncodeUint(uint64)
+	EncodeFloat(float64)
+	EncodeComplex(complex128)
+	EncodeBytes([]byte)
+
+	// EncodeList is called when a slice or array is encountered (except for a
+	// []byte, which is handled by EncodeBytes). Its argument is the length of the
+	// slice or array. The encoding algorithm will call the returned Encoder that
+	// many times to encode the successive values of the list. After each such call,
+	// ListIndex will be called with the index of the element just encoded.
+	//
+	// For example, []string{"a", "b"} will result in these calls:
+	//     enc2 := enc.EncodeList(2)
+	//     enc2.EncodeString("a")
+	//     enc2.ListIndex(0)
+	//     enc2.EncodeString("b")
+	//     enc2.ListIndex(1)
+	EncodeList(n int) Encoder
+	ListIndex(i int)
+
+	// EncodeMap is called when a map or struct is encountered. Its argument is the
```

Comment:
This is a bit misleading for structs, since it's actually only called for structs if `EncodeStruct` returns false.

I would consider just having this paragraph/example reference the map use case, and have a sentence at the end that says something like "EncodeMap is also called for structs if EncodeStruct returns false; the map then then consists of the public fields of the struct."

Reply:
Done.